README.md
# timeline-dl

Download Twitter timeline images

## Install

Run directly

```bash
npx timeline-dl
```

Install on system

```bash
npm i -g timeline-dl
%binName%
```

## Twitter

1. First create a new account. It can be blocked so don't use your main account!
2. Then use F12 to open Chrome/Firefox's developer tool, find cookie `auth_token` and copy the value.
3. Create a new folder and run `TWITTER_AUTH_TOKEN=xxx timeline-dl twitter KimKardashian`

## Options

### `--help`

Show help

### `--version`

Show version


CHANGELOG.md
# Changelog

## 1.1.0 - 2023-12-17

- YouTube video download

## 1.0.0 - 2023-12-17

- Twitter image download


src/utils/download.ts
import chalk from 'chalk';
import fse from 'fs-extra';
import https from 'https';

const { writeFile } = fse;

export function download(url: string, filename: string, retry = 5): Promise<void> {
  return new Promise((resolve, reject) => {
    console.log(url, chalk.blue('Start'));
    https
      .get(url, (res) => {
        const data: any[] = [];
        res
          .on('data', (chunk) => {
            data.push(chunk);
          })
          .on('end', () => {
            const buffer = Buffer.concat(data);
            if (buffer.byteLength > 1024) {
              console.log(url, chalk.green('Succeeded'));
              writeFile(filename, buffer);
              resolve();
            } else if (retry > 0) {
              console.log(url, chalk.yellow('Retry...'));
              download(url, filename, retry - 1)
                .then(resolve)
                .catch(reject);
            }
          });
      })
      .on('error', (err) => {
        console.log(url, 'error', err);
        if (retry > 0) {
          console.log(url, chalk.yellow('Retry...'));
          download(url, filename, retry - 1)
            .then(resolve)
            .catch(reject);
        } else {
          console.log(url, chalk.red('Failed'));
          reject();
        }
      });
  });
}


src/scrapers/youtube.ts
import fse from 'fs-extra';
import { launch } from 'puppeteer-core';
import { execute } from '../utils/execute';

const { readJson, writeJson, readdir } = fse;

const videoSelector = 'a[href^="/watch?v="]';

declare const location: any;
declare const document: any;
declare const window: any;

export interface TwitterOptions {
  headless: boolean;
  imageFormat: string;
}

export async function youtube(username: string | null, { headless, imageFormat }: TwitterOptions) {
  let config: any = {};
  try {
    config = await readJson('youtube.json', { throws: false });
  } catch (e) {
    //
  }
  const browser = await launch({
    executablePath: '/usr/bin/chromium',
    headless,
    defaultViewport: {
      width: 1280,
      height: 2560,
    },
  });

  if (!username) return;

  console.log(`Fetch @${username} https://youtube.com/@${username}/videos`);

  const startTime = new Date();
  const page = await browser.newPage();

  const existingFiles = await readdir(process.cwd());

  // accept cookies
  await page.setCookie({
    name: 'CONSENT',
    value: 'PENDING+369',
    domain: '.google.com',
    path: '/',
    expires: 1712538612.671163,
    httpOnly: false,
    secure: false,
    sameParty: false,
    sourceScheme: 'Secure',
    sourcePort: 443,
  });
  await page.setCookie({
    name: 'CONSENT',
    value: 'PENDING+206',
    domain: '.youtube.com',
    path: '/',
    expires: 1712538612.671163,
    httpOnly: false,
    secure: false,
    sameParty: false,
    sourceScheme: 'Secure',
    sourcePort: 443,
  });
  await page.setCookie({
    name: 'SOCS',
    value:
      'CAISOAgDEitib3FfaWRlbnRpdHlmcm9udGVuZHVpc2VydmVyXzIwMjMxMjEyLjA1X3AwGgV6aC1DTiACGgYIgPT4qwY',
    domain: '.youtube.com',
    path: '/',
    expires: 1712538612.671163,
    httpOnly: false,
    secure: true,
    sameParty: false,
    sourceScheme: 'Secure',
    sameSite: 'Lax',
    sourcePort: 443,
  });

  try {
    await page.goto(`https://youtube.com/@${username}/videos`, {
      waitUntil: 'load',
      timeout: 30 * 1000,
    });

    const doneList = new Set<string>();

    // eslint-disable-next-line no-constant-condition
    while (true) {
      const videos = await page.$$eval(videoSelector, (es) => {
        return es.map((e) => e.href.split('=')[1]);
      });

      for (const video of videos) {
        console.log(video);
        if (doneList.has(video)) {
          continue;
        }
        // video was already downloaded, skip
        if (existingFiles.some((f) => f.includes(`[${video}]`) && !f.endsWith('.part'))) {
          doneList.add(video);
          continue;
        }
        await execute(
          'yt-dlp --retries=infinite --retry-sleep=3 -o "%(upload_date)s %(title)s [%(id)s].%(ext)s" --restrict-filenames https://www.youtube.com/watch?v=' +
            video
        );
        doneList.add(video);
      }

      await page.keyboard.press('PageDown');
    }
    config.updatedAt = startTime.toISOString();
    await writeJson('youtube.json', config, { spaces: 2 });
  } catch (e) {
    console.error(e);
  }

  await browser.close();
}


src/scrapers/twitter.ts
import fse from 'fs-extra';
import { launch } from 'puppeteer-core';
import sanitize from 'sanitize-filename';
import { download } from '../utils/download';

const { readJson, writeJson } = fse;

const tweetSelector = 'article[data-testid="tweet"]';

declare const location: any;
declare const document: any;
declare const window: any;

export interface TwitterOptions {
  headless: boolean;
  imageFormat: string;
}

export async function twitter(username: string | null, { headless, imageFormat }: TwitterOptions) {
  let config: any = {};
  try {
    config = await readJson('twitter.json', { throws: false });
  } catch (e) {
    //
  }
  const browser = await launch({
    executablePath: '/usr/bin/chromium',
    headless,
    defaultViewport: {
      width: 1280,
      height: 2560,
    },
  });

  if (!username) return;

  console.log(`Fetch @${username} https://twitter.com/${username}/`);

  const startTime = new Date();
  const page = await browser.newPage();

  try {
    await page.setCookie({
      name: 'auth_token',
      value: process.env?.['TWITTER_AUTH_TOKEN'] || '',
      domain: '.twitter.com',
      path: '/',
      expires: 1712538612.671163,
      httpOnly: false,
      secure: false,
      sameParty: false,
      sourceScheme: 'Secure',
      sourcePort: 443,
    });

    await page.goto(`https://twitter.com/${username}/?lang=en`, {
      waitUntil: 'load',
      timeout: 30 * 1000,
    });

    const doneList = new Set<string>();

    // eslint-disable-next-line no-constant-condition
    while (true) {
      await page.waitForNetworkIdle({ timeout: 0 });

      const tweets = await page.$$eval(tweetSelector, (es) => {
        const username = location.pathname.split('/')[1];
        return es.map((e) => {
          const tweetText = e.querySelector('[data-testid="tweetText"]');
          // replace emoji image with emoji text
          Array.from(tweetText?.querySelectorAll('img') || []).forEach((img: any) => {
            const span = document.createElement('span');
            span.textContent = img.alt;
            img?.replaceWith(span);
          });
          const result = {
            link: e
              .querySelector(`a[href*="${username}/status/"]`)
              ?.href?.replace('/analytics', ''),
            text: tweetText?.textContent,
            dateTime: e.querySelector('time').dateTime,
            images: Array.from(e.querySelectorAll('img') || [])
              .map((img: any) => img.src)
              .filter((src) => src.startsWith('https://pbs.twimg.com/media/')),
          };
          return result;
        });
      });

      await Promise.all(
        tweets
          .filter(
            (tweet) =>
              tweet.images.length > 0 &&
              tweet.link &&
              !doneList.has(tweet.link) &&
              (!config.updatedAt ||
                new Date(config.updatedAt).getTime() < new Date(tweet.dateTime).getTime())
          )
          .map(async (tweet) => {
            if (tweet.images.length > 0) {
              await Promise.all(
                tweet.images.map(async (img, index) => {
                  const imgUrl =
                    img.substring(0, img.indexOf('?')) + `?format=${imageFormat}&name=4096x4096`;
                  const imgFileName = sanitize(
                    `${tweet.dateTime.substring(0, 10)} ${tweet.text || ''} (${
                      index + 1
                    }).${imageFormat}`
                  );
                  try {
                    await download(imgUrl, imgFileName);
                  } catch (e) {
                    //
                  }
                })
              );
            }
          })
      );

      const lastTweet = tweets.pop();

      if (
        config.updatedAt &&
        lastTweet &&
        new Date(config.updatedAt).getTime() > new Date(lastTweet.dateTime).getTime()
      ) {
        break;
      }

      const scrollStatus = await page.$eval('[data-testid="cellInnerDiv"]:last-child', (e) => {
        const rect = e.getBoundingClientRect();
        if (
          rect.top >= 0 &&
          rect.left >= 0 &&
          rect.bottom <= (window.innerHeight || document.documentElement.clientHeight) &&
          rect.right <= (window.innerWidth || document.documentElement.clientWidth)
        ) {
          return 'end';
        } else {
          return 'loading';
        }
      });

      if (scrollStatus === 'end') {
        break;
      } else {
        await page.keyboard.press('PageDown');
      }
      continue;
    }
    config.updatedAt = startTime.toISOString();
    await writeJson('twitter.json', config, { spaces: 2 });
  } catch (e) {
    console.error(e);
  }

  await browser.close();
}
